app/jobs/autoscaling.py

```diff
@@ -260,227 +61,313 @@ def get_instance_tags():
 
-
-def instances_in(asg):
-    return [item["InstanceId"] for item in asg["Instances"]]
-
-
-def find_tags_for_instances_in(asg, tag_list):
-    instance_ids = instances_in(asg)
-    return [tag for tag in tag_list if tag["ResourceId"] in instance_ids]
-
-
-def add_draining_tag(instance_ids, draining_tag):
-    cmd = "aws ec2 create-tags --resources {list_instances} --tags Key={draining_tag},Value={timestamp}"
-    cmd = cmd.format(list_instances=' '.join(instance_ids), draining_tag=draining_tag, timestamp=unixtime_now())
+def delete_tags_from_instances(instance_ids, tag_keys):
+    if not instance_ids or not tag_keys:
+        return
+    cmd = "aws ec2 delete-tags --resources {list_instances} --tags {list_tags}"
+    cmd = cmd.format(list_instances=' '.join(instance_ids), list_tags=' '.join(['Key='+tag for tag in tag_keys]))
     aws_command(cmd)
 
+def get_asg_list():
+    asg_json = aws_command("aws autoscaling describe-auto-scaling-groups")
+    asg_list = json.loads(asg_json).get('AutoScalingGroups', [])
+    return asg_list
 
-def remove_draining_tag(instance_ids, draining_tag):
-    delete_tags_from_instances(instance_ids, [draining_tag])
-
+def required_capacity(num_jobs):
+    if num_jobs == 0:
+        result = exactly(0)
+    elif num_jobs == 1:
+        # Often a test job or automatic benchmark job
+        result = exactly(4)
+    elif 1 < num_jobs <= 6:
+        result = exactly(8)
+    else:
+        # This means, as long as we're in this regime, a DesiredCapacity > 24 set manually by the operator
+        # will be respected and not scaled down until we exit this regime.
+        result = at_least(24)
+    return result
 
-def start_draining(asg_instance_name, draining_tag, num_instances):
-    instance_ids = instances_to_drain(asg_instance_name, draining_tag, num_instances)
-    if instance_ids:
-        print "Starting to drain the following instances: " + ",".join(instance_ids)
-        add_draining_tag(instance_ids, draining_tag)
+def autoscaling_update(my_num_jobs, my_environment="development",
+                       max_job_dispatch_lag_seconds=900, job_tag_prefix="RunningIDseqBatchJob_",
+                       job_tag_keep_alive_seconds=600, draining_tag="draining"):
+    '''
+    Plan and perform an update to the GSNAP/RAPSEARCH ASGs.
+    Note: Will not scale up for jobs originating in a development environment.
+          You can provision machines for development jobs by manually setting MinSize on the ASGs,
+          which is respected by this autoscaler.
+    '''
 
+    asg_list = get_asg_list()
+    tag_list = get_tag_list()
+    gsnap_ASG = ASG("gsnapl", my_environment, asg_list, tag_list, draining_tag, job_tag_prefix, max_job_dispatch_lag_seconds, job_tag_keep_alive_seconds)
+    rapsearch_ASG = ASG("rapsearch2", my_environment, asg_list, tag_list, draining_tag, job_tag_prefix, max_job_dispatch_lag_seconds, job_tag_keep_alive_seconds)
+    if not (gsnap_ASG.can_scale and rapsearch_ASG.can_scale):
```

Comment:
Is this still necessary?   If I understand correctly, the new rule is that development cannot scale anything;  and every other environment has exactly one autoscaler, which is permitted to scale it.   The reason this code was written originally was twofold:   development had to be prevented from scaling, but also there was the opportunity for a staging autoscaler to be allowed to scale the production ASGs in case the production autoscaler died.  This is too sophisticated and unnecessary now, I think.

Reply:
I mean this particular code can stay but the tags that previously determined can_scale should probably be retired.

Reply:
You're right -- I will remove

Reply:
based on subsequent discussion, leaving this in. It provides a mechanism to easily remove ASGs from the autoscaler's authority by deleting tags if things go wrong